src/ema/math.rs

```diff
@@ -0,0 +1,248 @@
+use crate::fraction;
+use crate::to_u128_wrapper;
+use crate::transcendental::saturating_powi_high_precision;
+use crate::types::{Balance, Fraction, Ratio};
+
+use num_traits::{One, Zero};
+use primitive_types::{U128, U256, U512};
+
+/// EmaPrice is a rational number represented by a `u128` for both numerator and denominator.
+pub type EmaPrice = Ratio;
+
+/// Calculate the iterated exponential moving average for the given prices.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_price_ema(iterations: u32, prev: EmaPrice, incoming: EmaPrice, smoothing: Fraction) -> EmaPrice {
+    price_weighted_average(prev, incoming, exp_smoothing(smoothing, iterations))
+}
+
+/// Calculate the iterated exponential moving average for the given balances.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_balance_ema(iterations: u32, prev: Balance, incoming: Balance, smoothing: Fraction) -> Balance {
+    balance_weighted_average(prev, incoming, exp_smoothing(smoothing, iterations))
+}
+
+/// Calculate the iterated exponential moving average for the given volumes.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_volume_ema(
+    iterations: u32,
+    prev: (Balance, Balance, Balance, Balance),
+    incoming: (Balance, Balance, Balance, Balance),
+    smoothing: Fraction,
+) -> (Balance, Balance, Balance, Balance) {
+    volume_weighted_average(prev, incoming, exp_smoothing(smoothing, iterations))
```

Comment:
I don't think this is correct, unlike with balance and price oracles we should assume that volume = 0 in blocks where this oracle was not updated.

Reply:
I think this belongs in the pallet where it is implemented and tested. See [this line here](https://github.com/galacticcouncil/warehouse/pull/76/files#diff-d3c5bdb6ebc4793ff959e9c94262acfdb51325d81253a0861b9c95b8e0d35894R357)

Reply:
Argh github thwarting me, here is [a permalink](https://github.com/galacticcouncil/warehouse/blob/bfd8f14a2128f136b0286683e7a2dc4b6e8688c9/ema-oracle/src/tests/mod.rs#L509)

Reply:
I haven't confirmed that the current implementation reaches the correct result due to how much is implemented in warehouse, though I could believe that it does.

However I would still prefer to have the correct math for multi-block update of volume oracles all in the math package. The math for multi-block update for the volume oracles is actually simpler than for the other oracles, so I would prefer to just directly use this correct & simpler math.

I've specified the correct math here: https://www.notion.so/Lazy-Evaluation-of-EMA-volume-oracles-7ed24358282d41b1ae3578121dd610cf